arrows/vxl/average_frames.cxx

```diff
@@ -0,0 +1,722 @@
+      {
+        throw std::runtime_error( "Invalid exponential averaging coefficient!" );
+      }
+
+      if( is_byte )
+      {
+        byte_averager.reset(
+          new exponential_frame_averager< vxl_byte >( round, exp_weight ) );
+      }
+      else
+      {
+        float_averager.reset(
+          new exponential_frame_averager< double >( round, exp_weight ) );   
+      }
+    }
+    else
+    {
+      throw std::runtime_error( "Invalid averaging type!" );
+    }
+  }
+};
+
+// --------------------------------------------------------------------------------------
+average_frames
+::average_frames()
+: d( new priv() )
+{
+  attach_logger( "arrows.vxl.average_frames" );
+}
+
+average_frames
+::~average_frames()
+{
+}
+
+vital::config_block_sptr
+average_frames
+::get_configuration() const
+{
+  // get base config from base class
+  vital::config_block_sptr config = algorithm::get_configuration();
+
+  config->set_value( "type", d->type,
+    "Operating mode of this filter, possible values: cumulative, window, "
+    "or exponential." );
+  config->set_value( "window_size", d->window_size,
+    "The window size if computing a windowed moving average." );
+  config->set_value( "exp_weight", d->exp_weight,
+    "Exponential averaging coefficient if computing an exp average." );
+  config->set_value( "round", d->round,
+    "Should we spend a little extra time rounding when possible?" );
+  config->set_value( "output_variance", d->output_variance,
+    "If set, will compute an estimated variance for each pixel which "
+    "will be outputted as either a double-precision or byte image." );
+
+  return config;
+}
+
+void
+average_frames
+::set_configuration( vital::config_block_sptr in_config )
+{
+  // Starting with our generated vital::config_block to ensure that assumed values
+  // are present. An alternative is to check for key presence before performing a
+  // get_value() call.
+  vital::config_block_sptr config = this->get_configuration();
+  config->merge_config( in_config );
+
+  // Settings for averaging
+  d->type = config->get_value< std::string >( "type" );
+  d->window_size = config->get_value< unsigned >( "window_size" );
+  d->exp_weight = config->get_value< double >( "exp_weight" );
+  d->round = config->get_value< bool >( "round" );
+  d->output_variance = config->get_value< bool >( "output_variance" );
+}
+
+bool
+average_frames
+::check_configuration( vital::config_block_sptr config ) const
+{
+  return true;
+}
+
+// Perform stitch operation
+kwiver::vital::image_container_sptr
+average_frames
+::filter( kwiver::vital::image_container_sptr image_data )
+{
+  // Perform Basic Validation
+  if( !image_data )
+  {
+    return image_data;
+  }
+
+  // Get input image
+  vil_image_view_base_sptr view =
+    vxl::image_container::vital_to_vxl( image_data->get_image() );
+
+  // Perform different actions based on input type
+#define HANDLE_CASE(T)                                                 \
```

Comment:
Is this the preferred of dealing with different VXL parameterizations in the underlying image data? I'm implementing similar filters and have not seen a better way, but wanted to make sure I'm not missing something.

Reply:
In general, yes, VXL uses this `DO_CASE` or `HANDLE_CASE` macro pattern to deal with dispatching to different code based on dynamic type.  However, in this case it's not done very cleanly.  Usually the macro is just a couple lines of code to find the type and call a templated function.   That's sort of what we have here except everything after the `vil_convert_cast` does not depend on the type, so it doesn't need to be written out in the macro.  It would be better to call some helper function to do that stuff because 1. it reduces code bloat and 2. it would probably be easier to debug.

Reply:
Thanks, that's really helpful. I think this has been updated appropriately.